### PR TITLE
raftstore: apply snapshot asynchronously

### PR DESCRIPTION
**Pull request by BusyJay**
refer: https://github.com/pingcap/tikv/issues/594

@ngaut @siddontang @disksing PTAL

### CODE REVIEW

File: src/raftstore/store/peer_storage.rs [hunk 1]
```diff
@@ -53,7 +56,8 @@ pub struct PeerStorage {
     pub region: metapb::Region,
     pub local_state: RaftLocalState,
 
-    pub snap_state: SnapState,
+    snap_state: Arc<RwLock<SnapState>>,
+    snap_sched: Scheduler<SnapTask>,
     snap_mgr: SnapManager,
```
Comment (siddontang) [inline src/raftstore/store/peer_storage.rs]: Why not move snap worker into SnapManager. So we can only use snap_mgr all places.

File: src/raftstore/store/worker/snap.rs [hunk 2]
```diff
@@ -49,47 +60,101 @@ quick_error! {
...
+        let timer = Instant::now();
+        // Write the snapshot into the region.
+        loop {
+            // TODO: avoid too many allocation
+            let key = box_try!(reader.decode_compact_bytes());
+            if key.is_empty() {
+                break;
+            }
+            let value = box_try!(reader.decode_compact_bytes());
+            box_try!(self.db.put(&key, &value));
```
Comment (siddontang) [inline src/raftstore/store/worker/snap.rs]: is using a batch for multiply paris faster than using put every time?

/cc @zhangjinpeng1987

File: src/raftstore/store/peer.rs [hunk 3]
```diff
@@ -335,6 +344,10 @@ impl Peer {
         let apply_result = try!(self.storage.wl().handle_raft_ready(&ready));
 
         for msg in &ready.messages {
+            if msg.get_msg_type() == MessageType::MsgRequestVote && is_applying {
```
Comment (siddontang) [inline src/raftstore/store/peer.rs]: is it ok for handing ready when applying?
/cc @xiang90 

Btw, can we need ignoring other peers's message too when applying?

Comment (xiang90) [inline src/raftstore/store/peer.rs]: the main thing to keep in mind is that:

raft state and application state are separate by design. Snapshot contains two part:

1. raft state

2. application state.

When you start to restore raft state, raft state machine needs to be paused. After you restore raft state from snapshot, you can restore application state whenever you want as long as you buffer the incoming entries. After you finish apply the snapshot, you start to apply the buffered entries.

File: src/raftstore/coprocessor/dispatcher.rs [hunk 4]
```diff
@@ -214,7 +215,9 @@ mod test {
 
     fn new_peer_storage(path: &TempDir) -> PeerStorage {
         let engine = new_engine(path.path().to_str().unwrap()).unwrap();
-        PeerStorage::new(engine, &Region::new(), store::new_snap_mgr("")).unwrap()
+        let worker = Worker::new("");
```
Comment (ngaut) [inline src/raftstore/coprocessor/dispatcher.rs]: Please set  name of the worker.

File: src/raftstore/coprocessor/region_snapshot.rs [hunk 5]
```diff
@@ -222,7 +223,9 @@ mod tests {
     }
 
     fn new_peer_storage(engine: Arc<DB>, r: &Region) -> PeerStorage {
-        PeerStorage::new(engine, r, store::new_snap_mgr("")).unwrap()
+        let worker = Worker::new("");
```
Comment (ngaut) [inline src/raftstore/coprocessor/region_snapshot.rs]: Ditto.

File: src/raftstore/coprocessor/split_observer.rs [hunk 6]
```diff
@@ -108,19 +108,22 @@ mod test {
...
     use util::codec::{datum, table, Datum};
     use util::codec::number::NumberEncoder;
     use util::codec::bytes::encode_bytes;
+    use util::worker::Worker;
     use byteorder::{BigEndian, WriteBytesExt};
 
     fn new_peer_storage(path: &TempDir) -> PeerStorage {
         let engine = new_engine(path.path().to_str().unwrap()).unwrap();
-        PeerStorage::new(engine, &Region::new(), store::new_snap_mgr("")).unwrap()
+        let worker = Worker::new("");
```
Comment (ngaut) [inline src/raftstore/coprocessor/split_observer.rs]: Ditto

File: src/raftstore/coprocessor/split_observer.rs [hunk 7]
```diff
@@ -162,7 +165,9 @@ mod test {
         r.set_id(10);
         r.set_start_key(region_start_key);
 
-        let ps = PeerStorage::new(engine, &r, store::new_snap_mgr("")).unwrap();
+        let worker = Worker::new("");
```
Comment (ngaut) [inline src/raftstore/coprocessor/split_observer.rs]: Abstract a new method to get scheduler reference.

File: src/raftstore/store/worker/snap.rs (hunk 2 above)
Comment (zhangjinpeng1987) [inline src/raftstore/store/worker/snap.rs]: yes

File: src/raftstore/store/peer_storage.rs [hunk 8]
```diff
@@ -494,6 +482,11 @@ impl PeerStorage {
         self.local_state = ctx.local_state;
         // If we apply snapshot ok, we should update some infos like applied index too.
         if let Some(res) = apply_snap_res {
+            self.snap_state = SnapState::Applying;
+            self.snap.take();
+            let task = SnapTask::Apply { region_id: region_id };
+            // TODO: gracefully remove region instead.
+            self.snap_sched.schedule(task).expect("snap apply job should not failed");
```
Comment (siddontang) [inline src/raftstore/store/peer_storage.rs]: should not fail

**Comment by siddontang** [comment]
LGTM

File: src/raftstore/coprocessor/region_snapshot.rs [hunk 9]
```diff
@@ -222,7 +223,7 @@ mod tests {
     }
 
     fn new_peer_storage(engine: Arc<DB>, r: &Region) -> PeerStorage {
-        PeerStorage::new(engine, r, store::new_snap_mgr("")).unwrap()
+        PeerStorage::new(engine, r, worker::dead_scheduler()).unwrap()
```
Comment (ngaut) [inline src/raftstore/coprocessor/region_snapshot.rs]: s/dead_scheduler/dummy_scheduler

File: src/raftstore/store/peer.rs [hunk 10]
```diff
@@ -328,16 +339,30 @@ impl Peer {
                self.peer,
                self.region_id);
 
-        let ready = self.raft_group.ready();
+        let mut ready = self.raft_group.ready();
+
```
Comment (ngaut) [inline src/raftstore/store/peer.rs]: Remove blank line.

File: src/raftstore/store/peer.rs [hunk 11]
```diff
@@ -328,16 +339,30 @@ impl Peer {
...
+        let is_applying = self.storage.rl().is_applying_snap();
+
```
Comment (ngaut) [inline src/raftstore/store/peer.rs]: Ditto.

File: src/raftstore/store/peer.rs [hunk 12]
```diff
@@ -328,16 +339,30 @@ impl Peer {
...
+            // skip apply and snapshot
+            ready.committed_entries = vec![];
+            ready.snapshot = RaftSnapshot::new();
+        }
 
         let t = SlowTimer::new();
 
         self.send_ready_metric(&ready);
 
+        // The leader can write to its disk in parallel with replicating to the followers.
```
Comment (ngaut) [inline src/raftstore/store/peer.rs]: The leader can write to disk and replicate to the followers concurrently

File: src/raftstore/store/peer.rs [hunk 13]
```diff
@@ -354,6 +379,11 @@ impl Peer {
                   ready.hs.is_some(),
                   t.elapsed());
 
+        if is_applying {
+            // remove hard state to let raft not change apply index.
```
Comment (ngaut) [inline src/raftstore/store/peer.rs]: remove hard state so raft won't change the apply index.

File: src/raftstore/store/worker/snap.rs [hunk 14]
```diff
@@ -48,48 +62,144 @@ quick_error! {
...
+            if key.is_empty() {
+                box_try!(self.db.write(wb));
+                break;
+            }
+            batch_size += key.len();
+            let value = box_try!(reader.decode_compact_bytes());
+            batch_size += value.len();
+            box_try!(wb.put(&key, &value));
+            if batch_size > BATCH_SIZE {
+                box_try!(self.db.write(wb));
```
Comment (ngaut) [inline src/raftstore/store/worker/snap.rs]: write_without_wal ?

File: tests/raftstore/node.rs [hunk 15]
```diff
@@ -130,22 +130,35 @@ impl Simulator for NodeCluster {
...
-        let snap_mgr = store::new_snap_mgr(tmp.path().to_str().unwrap());
-
         let mut event_loop = create_event_loop(&cfg.store_cfg).unwrap();
         let simulate_trans = SimulateTransport::new(self.trans.clone());
         let trans = Arc::new(RwLock::new(simulate_trans));
         let mut node = Node::new(&mut event_loop, &cfg, self.pd_client.clone());
 
+        let (snap_mgr, tmp) = if node_id == 0 ||
+                                 !self.trans.rl().snap_paths.contains_key(&node_id) {
+            let tmp = TempDir::new("test_luster").unwrap();
```
Comment (ngaut) [inline tests/raftstore/node.rs]: s/test_luster/test_cluster/

File: tests/raftstore/test_snap.rs [hunk 16]
```diff
@@ -143,6 +143,8 @@ fn test_snap_gc<T: Simulator>(cluster: &mut Cluster<T>) {
             panic!("snap files is still not empty: {:?}", snap_files);
         }
         tried_cnt += 1;
+        // trigger compact log.
```
Comment (ngaut) [inline tests/raftstore/test_snap.rs]: s/compact log/log compaction/

**Comment by ngaut** [comment]
Rest LGTM

**Comment by ngaut** [comment]
LGTM

**Pull request merged**